src/libraries/Common/src/System/Security/Cryptography/SlhDsa.cs

```diff
@@ -274,6 +270,211 @@ public bool VerifyData(byte[] data, byte[] signature, byte[]? context = default)
+        /// </exception>
+        /// <exception cref="CryptographicException">
+        ///   <para><paramref name="hashAlgorithmOid"/> is not a well-formed OID.</para>
+        ///   <para>-or-</para>
+        ///   <para><paramref name="hashAlgorithmOid"/> is a well-known algorithm and <paramref name="hash"/> does not have the expected length.</para>
+        ///   <para>-or-</para>
+        ///   <para>The instance represents only a public key.</para>
+        ///   <para>-or-</para>
+        ///   <para>An error occurred while signing the hash.</para>
+        /// </exception>
+        /// <remarks>
+        ///   A <see langword="null" /> context is treated as empty.
+        /// </remarks>
+        public byte[] SignPreHash(byte[] hash, string hashAlgorithmOid, byte[]? context = default)
+        {
+            ArgumentNullException.ThrowIfNull(hash);
+            ArgumentNullException.ThrowIfNull(hashAlgorithmOid);
+
+            byte[] destination = new byte[Algorithm.SignatureSizeInBytes];
+            SignPreHash(new ReadOnlySpan<byte>(hash), destination.AsSpan(), hashAlgorithmOid, new ReadOnlySpan<byte>(context));
+            return destination;
+        }
+
+        /// <summary>
+        ///   Verifies that the specified FIPS 205 pre-hash signature is valid for this key and the provided hash.
+        /// </summary>
+        /// <param name="hash">
+        ///   The hash to verify.
+        /// </param>
+        /// <param name="signature">
+        ///   The signature to verify.
+        /// </param>
+        /// <param name="hashAlgorithmOid">
+        ///   The OID of the hash algorithm used to create the hash.
+        /// </param>
+        /// <param name="context">
+        ///   The context value which was provided during signing.
+        ///   The default value is an empty buffer.
+        /// </param>
+        /// <returns>
+        ///   <see langword="true"/> if the signature validates the hash; otherwise, <see langword="false"/>.
+        /// </returns>
+        /// <exception cref="ArgumentOutOfRangeException">
+        ///   <paramref name="context"/> has a <see cref="ReadOnlySpan{T}.Length"/> in excess of
+        ///   255 bytes.
+        /// </exception>
+        /// <exception cref="ObjectDisposedException">
+        ///   This instance has been disposed.
+        /// </exception>
+        /// <exception cref="CryptographicException">
+        ///   <para><paramref name="hashAlgorithmOid"/> is not a well-formed OID.</para>
+        ///   <para>-or-</para>
+        ///   <para><paramref name="hashAlgorithmOid"/> is a well-known algorithm and <paramref name="hash"/> does not have the expected length.</para>
+        ///   <para>-or-</para>
+        ///   <para>An error occurred while verifying the hash.</para>
+        /// </exception>
+        public bool VerifyPreHash(ReadOnlySpan<byte> hash, ReadOnlySpan<byte> signature, ReadOnlySpan<char> hashAlgorithmOid, ReadOnlySpan<byte> context = default)
```

Comment:
Do we have a strong use case for ROS-char hashAlgorithmOid over string hashAlgorithmOid?  Windows could, for example, end up exposing it as pszHashOid, which means we need the guaranteed null terminator.

We should probably start based on string (here and sign and the protected members), and weaken it to ROS-char later after seeing what Windows does.

Reply:
Yep, docs say null terminated: https://learn.microsoft.com/en-us/windows/win32/seccng/bcrypt/ns-bcrypt-bcrypt_pqdsa_padding_info#syntax